[LC-557] Keep order after penalty

zsaladin opened the pull request:
Keep order after penalty
- Add `build_next_leader`
- Add `build_next_reps_hash`
- maintain leader and made_block_count when leader no need to change in penalty
- keep order after penalty will work on block version `0.4`

zsaladin committed:
[LC-557] base of keep order after penalty

 - set `next_reps` into `block_builder` to verify `next_leader`
 - Keep reps order after panalty
 - Fix that invoke result does not be returned

zsaladin commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -29,9 +29,19 @@
+
+    def verify_loosely(self, block: 'Block', prev_block: 'Block', blockchain=None, **kwargs):
+        kwargs.pop('next_leader', None)
+        super().verify(block, prev_block, blockchain, **kwargs)
```
verify -> verify_loosely

yakkle commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -118,6 +127,19 @@ def verify_invoke(self, builder: 'BlockBuilder', block: 'Block', prev_block: 'Bl
                                      f"Expected({new_block.header.next_reps_hash}).")
             self._handle_exception(exception)
 
+        next_leader = new_block.header.next_leader
```
The `next_leader` which get from parameter is not use and assigned from `new_block`.

zsaladin replied:
This code must be excluded.

yakkle commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -158,18 +173,6 @@ def verify_leader_votes(self, block: 'Block', prev_block: 'Block',  reps: Sequen
-                                         f"Leader({block.header.peer_id.hex_xx()}), "
-                                         f"Expected({prev_block_header.next_leader.hex_xx()}).\n "
-                                         f"LeaderVotes({body.leader_votes}")
-                self._handle_exception(exception)
```
why this `else phrase` deleted?

zsaladin replied:
Oh, it should not be deleted.

zsaladin commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -118,6 +127,19 @@ def verify_invoke(self, builder: 'BlockBuilder', block: 'Block', prev_block: 'Bl
+                                     f"NextLeader({header.next_leader}), "
+                                     f"Expected({next_leader}).")
+            self._handle_exception(exception)
+
```
These codes move to line 117

yakkle replied:
I fixed. please review again.

zsaladin commented on loopchain/blockchain/blocks/v0_3/block_builder.py:
```diff
@@ -128,20 +131,45 @@ def _build_receipts_hash(self):
+            next_index = next_index if next_index < len(self.next_reps) else 0
+            return self.next_reps[next_index]
+        else:
+            return self.next_leader
```
Right

yakkle committed:
[LC-557] maintain leader and made_block_count

 - maintain leader and made_block_count when leader no need to change in penalty

yakkle committed:
[LC-557] fixed error BlockHeader test

 - change PrepChangedReason to NextRepsChangeReason
 - update test_block_v0_3 verify parameters

metalg0su commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1101,11 +1117,19 @@ def genesis_invoke(self, block: Block, prev_block_ = None) -> ('Block', dict):
+            return NextRepsChangeReason.Penalty
+
+        return NextRepsChangeReason.NoChange
+
```
Is there any reason that this method belongs to `BlockChain` class?

yakkle replied:
Nope. It have to move to proper location.

hyerim-kim commented on loopchain/peer/block_manager.py:
```diff
@@ -855,7 +858,7 @@ def leader_complain(self):
 
     def vote_unconfirmed_block(self, block: Block, round_: int, is_validated):
-        logging.debug(f"block_manager:vote_unconfirmed_block ({self.channel_name}/{is_validated})")
+        logging.debug(f"vote_unconfirmed_block() ({self.channel_name}/{is_validated})")
```
Empty brackets look like unnecessary.

hyerim-kim replied:
I got it, they mean a method.

hyerim-kim commented on loopchain/blockchain/blockchain.py:
```diff
@@ -130,13 +130,29 @@ def _increase_made_block_count(self, block: Block) -> None:
     def reset_leader_made_block_count(self):
-        self.__made_block_counter.clear()
+        """Clear all made_block_counter
+        FIXME : is func_name match to behavior?
```
I think appropriates enough.  That means "Reset `leader_made_block_count`.

hyerim-kim commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1101,11 +1117,19 @@ def genesis_invoke(self, block: Block, prev_block_ = None) -> ('Block', dict):
+            return NextRepsChangeReason.Penalty
+
+        return NextRepsChangeReason.NoChange
+
```
```suggestion
class NextRepsChangeReason(IntEnum):
    NoChange = -1
    TermEnd = 0
    Penalty = 1

    @classmethod
    def get_member(cls, value: Union[str, int]):
        if isinstance(value, str):
            int_value = int(value, 0)
        elif isinstance(value, int):
            int_value = value
        else:
            # return cls.NoChange or
            raise TypeError

        for member in list(cls._member_map_.values()):
            if member == int_value:
                return member
```
And use like this
```suggestion
block_builder.next_reps_change_reason = NextRepsChangeReason.get_item(next_prep["state"])
```
instead of 
```
block_builder.next_reps_change_reason = self._convert_reps_change_reason(next_prep["state"])
```

yakkle replied:
updated!

yakkle committed:
[LC-557] separate penalty from block version 0.3

 - skip verify unconfirmed_block.next_leader and expected next leader
 - keep order after penalty will work on block version 0.4

windies21 commented on loopchain/baseservice/node_subscriber.py:
```diff
@@ -31,6 +31,7 @@
 from loopchain.baseservice import ObjectManager, TimerService, Timer
 from loopchain.blockchain import AnnounceNewBlockError
 from loopchain.blockchain.blocks import BlockSerializer, BlockVerifier
+from loopchain.blockchain.types import ExternalAddress
```
Is it needed?

yakkle replied:
removed!

hyerim-kim committed:
[LC-557] Add logic to initialize `leader_made_block_count`

 - reset made_block_count when `Watch` is switched role to
   replace a `Voter` that got the penalty to `Voter`.
 - Add some logs.

zsaladin commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1131,6 +1131,62 @@ def genesis_invoke(self, block: Block, prev_block_=None) -> Tuple[Block, dict]:
+
+            next_leader = None  # to rebuild next_leader
+            block_builder.next_reps_change_reason = change_reason
+            block_builder.is_max_made_block_count = self.made_block_count_reached_max(_block)
```
`is_max_made_block_count` has been removed out of `BlockBuilder`

yakkle commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1131,6 +1131,62 @@ def genesis_invoke(self, block: Block, prev_block_=None) -> Tuple[Block, dict]:
         self.__invoke_results[new_block.header.hash] = (tx_receipts, None)
         return new_block, tx_receipts
 
+    def _process_next_prep(self, _block: Block, block_builder: BlockBuilder, next_prep: dict):
```
👌
